Ipset input plugin

sajoupa opened the pull request:
### Required for all PRs:

- [X] Signed [CLA](https://influxdata.com/community/cla/).
- [X] Associated README.md updated.
- [X] Has appropriate unit tests.

The ipset input plugin allows collecting metrics from Linux ipset counters.

danielnelson commented on plugins/inputs/ipset/README.md:
```diff
@@ -0,0 +1,59 @@
+There are 3 ways to grant telegraf the right to run ipset:
+* Run as root (strongly discouraged)
+* Use sudo
+* Configure systemd to run telegraf with CAP_NET_ADMIN and CAP_NET_RAW capabilities.
```
Are you sure this will work?  Are capabilities inherited?

danielnelson commented on plugins/inputs/ipset/ipset.go:
```diff
@@ -0,0 +1,106 @@
+  ## set show_all_sets = true to gather them all.
+  show_all_sets = false
+  ## Adjust your sudo settings appropriately if using this option ("sudo ipset save")
+  ## TODO: can we replace this with systemd privileges ? CAP_NET_ADMIN should DTRT
```
I guess this answers my question above, if we don't know if this will work don't add it.

danielnelson commented on plugins/inputs/ipset/ipset.go:
```diff
@@ -0,0 +1,106 @@
+	}
+
+	lines := strings.Split(list, "\n")
+	for _, line := range lines {
```
Use an bufio.Scanner to iterate over lines

danielnelson commented on plugins/inputs/ipset/ipset.go:
```diff
@@ -0,0 +1,106 @@
+			continue
+		}
+
+		data := strings.Split(line, " ")
```
`strings.Fields(line)`

danielnelson commented on plugins/inputs/ipset/ipset.go:
```diff
@@ -0,0 +1,106 @@
+		}
+
+		data := strings.Split(line, " ")
+		if data[0] == "add" && (data[4] != "0" || ips.ShowAllSets == true) {
```
Will panic if not enough fields, check for the correct number of splits.

danielnelson commented on plugins/inputs/ipset/ipset.go:
```diff
@@ -0,0 +1,106 @@
+	}
+	args = append(args, "save")
+
+	cmd := exec.Command(cmdName, args...)
```
Use `internal.RunTimeout` with an appropriate timeout, you may want to make it configurable.

Code example: https://github.com/influxdata/telegraf/blob/master/plugins/inputs/unbound/unbound.go#L69

danielnelson commented on plugins/inputs/ipset/ipset.go:
```diff
@@ -0,0 +1,106 @@
+	return string(out), err
+}
+
+type setLister func() (string, error)
```
Define this above with the struct

danielnelson commented on plugins/inputs/ipset/ipset_nocompile.go:
```diff
@@ -0,0 +1,3 @@
+// +build !linux
```
You can remove all of the build flags, it won't hurt to allow this to be built on all platforms.

danielnelson commented on plugins/inputs/ipset/ipset_test.go:
```diff
@@ -0,0 +1,119 @@
+	}
+
+	for i, tt := range tests {
+		i++
```
Use subtests like: https://github.com/influxdata/telegraf/blob/master/plugins/inputs/docker/docker_test.go#L392

danielnelson commented on plugins/inputs/ipset/ipset_test.go:
```diff
@@ -0,0 +1,119 @@
+		fields [][]map[string]interface{}
+		err    error
+	}{
+		{ // 1 - 0 sets => no results
```
Add a name field to the test struct and set the names based on these comments using subtests

danielnelson commented on plugins/inputs/ipset/README.md:
```diff
@@ -0,0 +1,59 @@
+  [[inputs.ipset]]
+    ## By default, we only show sets which have already matched at least 1 packet.
+    ## set show_all_sets = true to gather them all.
+    show_all_sets = false
```
I would call this `include_unmatched_sets`

sajoupa commented:
Thanks @danielnelson for the review.
The commit I just pushed addresses all the issues you pointed.
Also, I can confirm that systemd capabilities allow the service to run as intended, with an unprivileged user.

danielnelson commented on plugins/inputs/ipset/ipset.go:
```diff
@@ -0,0 +1,126 @@
+  include_unmatched_sets = false
+  ## Adjust your sudo settings appropriately if using this option ("sudo ipset save")
+  use_sudo = false
+  ## The default timeout of 1s for ss execution can be overridden here:
```
Replace ss with ipset, here and also remember to update the same in the README

sajoupa replied:
Oops! Fixed.

danielnelson commented:
What about this comment: https://github.com/influxdata/telegraf/pull/3346#discussion_r160807614

sajoupa commented:
@danielnelson Hi, I addressed https://github.com/influxdata/telegraf/pull/3346#discussion_r160807614 in https://github.com/sajoupa/telegraf/commit/72430de9e580eaf208aeb3d8785ef635162e2e15 .

danielnelson commented:
Ah, sorry was confused by the comment not collapsing.  I merged this in and it will be included in 1.6, thanks!
